src/node_serdes.cc

```diff
@@ -0,0 +1,457 @@
+}
+
+void SerializerContext::ReleaseBuffer(const FunctionCallbackInfo<Value>& args) {
+  SerializerContext* ctx;
+  ASSIGN_OR_RETURN_UNWRAP(&ctx, args.Holder());
+
+  std::vector<uint8_t> ret = ctx->serializer_.ReleaseBuffer();
+  auto buf = Buffer::Copy(ctx->env(),
+                          reinterpret_cast<const char*>(ret.data()),
+                          ret.size());
+
+  if (!buf.IsEmpty()) {
+    args.GetReturnValue().Set(buf.ToLocalChecked());
+  }
+}
+
+void SerializerContext::TransferArrayBuffer(
+    const FunctionCallbackInfo<Value>& args) {
+  SerializerContext* ctx;
+  ASSIGN_OR_RETURN_UNWRAP(&ctx, args.Holder());
+
+  uint32_t id = args[0]->Uint32Value();
+
+  if (args[1]->IsArrayBuffer()) {
+    Local<ArrayBuffer> ab = args[1].As<ArrayBuffer>();
+    ctx->serializer_.TransferArrayBuffer(id, ab);
+    return;
+  }
+
+  if (args[1]->IsSharedArrayBuffer()) {
+    Local<SharedArrayBuffer> sab = args[1].As<SharedArrayBuffer>();
+    ctx->serializer_.TransferSharedArrayBuffer(id, sab);
+    return;
+  }
+
+  return ctx->env()->ThrowTypeError("arrayBuffer must be an ArrayBuffer or "
+                                    "SharedArrayBuffer");
+}
+
+void SerializerContext::WriteUint32(const FunctionCallbackInfo<Value>& args) {
+  SerializerContext* ctx;
+  ASSIGN_OR_RETURN_UNWRAP(&ctx, args.Holder());
+
+  ctx->serializer_.WriteUint32(args[0]->Uint32Value());
+}
+
+void SerializerContext::WriteUint64(const FunctionCallbackInfo<Value>& args) {
+  SerializerContext* ctx;
+  ASSIGN_OR_RETURN_UNWRAP(&ctx, args.Holder());
+
+  uint64_t hi = args[0]->Uint32Value();
+  uint64_t lo = args[1]->Uint32Value();
+  ctx->serializer_.WriteUint64((hi << 32) | lo);
+}
+
+void SerializerContext::WriteDouble(const FunctionCallbackInfo<Value>& args) {
+  SerializerContext* ctx;
+  ASSIGN_OR_RETURN_UNWRAP(&ctx, args.Holder());
+
+  ctx->serializer_.WriteDouble(args[0]->NumberValue());
+}
+
+void SerializerContext::WriteRawBytes(const FunctionCallbackInfo<Value>& args) {
+  SerializerContext* ctx;
+  ASSIGN_OR_RETURN_UNWRAP(&ctx, args.Holder());
+
+  if (!args[0]->IsUint8Array()) {
+    return ctx->env()->ThrowTypeError("source must be a Uint8Array");
+  }
+
+  ctx->serializer_.WriteRawBytes(Buffer::Data(args[0]),
+                                 Buffer::Length(args[0]));
+}
+
+DeserializerContext::DeserializerContext(Environment* env,
+                                         Local<Object> wrap,
+                                         Local<Value> buffer,
+                                         Local<Value> readHostObject)
+  : BaseObject(env, wrap),
+    data_(reinterpret_cast<const uint8_t*>(Buffer::Data(buffer))),
+    length_(Buffer::Length(buffer)),
+    deserializer_(env->isolate(), data_, length_, this) {
+  object()->Set(env->context(), env->buffer_string(), buffer);
+  object()->Set(env->context(), env->read_host_object_string(), readHostObject);
+  deserializer_.SetSupportsLegacyWireFormat(true);
```

Comment:
Maybe explain in a comment that otherwise it always throws an "invalid or unsupported version" exception.